wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserNameValidator.java

```diff
@@ -18,28 +25,53 @@
  * @author Mihail Kuznyetsov
  */
 public class UserNameValidator {
+    private static final Logger LOG = LoggerFactory.getLogger(UserNameValidator.class);
+
     private static final Pattern ILLEGAL_USERNAME_CHARACTERS = Pattern.compile("[^a-zA-Z0-9]");
     private static final Pattern VALID_USERNAME              = Pattern.compile("^[a-zA-Z0-9]*");
 
+    private final UserManager userManager;
+
+    @Inject
+    public UserNameValidator (UserManager userManager) {
+        this.userManager = userManager;
+    }
+
     /**
      * Validate name, if it doesn't contain illegal characters
      *
      * @param name
      *        username
      * @return true if valid name, false otherwise
      */
-    public static boolean isValidUserName(String name) {
+    public boolean isValidUserName(String name) {
         return VALID_USERNAME.matcher(name).matches();
     }
 
     /**
      * Remove illegal characters from username, to make it URL-friendly.
+     * If all characters are illegal, return username.
+     * Also ensures username is unique, if not, adds digits to it's end.
      *
      * @param name
      *        username
      * @return username without illegal characters
      */
-    public static String normalizeUserName(String name) {
-        return ILLEGAL_USERNAME_CHARACTERS.matcher(name).replaceAll("");
+    public String normalizeUserName(String name) {
+        String normalized = ILLEGAL_USERNAME_CHARACTERS.matcher(name).replaceAll("");
+        String candidate = normalized.isEmpty() ? NameGenerator.generate("username", 4) : normalized;
+
+        int i = 1;
+        try {
+            while (true) {
+                userManager.getByName(candidate);
+                candidate = normalized.isEmpty() ? NameGenerator.generate("username", 4) : normalized + String.valueOf(i++);
+            }
+        } catch (NotFoundException e) {
+            return candidate;
+        } catch (ServerException e) {
+            LOG.warn("Error occured during username normalization", e);
```

Comment:
Maybe we should throw exception here instead of returning potential existing name